Update Identify docs

Lior539 opened the pull request:
This is an early draft of an update to the identify docs. For now, I'm just looking for feedback regarding technical accuracy, and if I've missed anything important. 

I still need to do other important things, like unify how we talk about it across the rest of docs

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,382 @@
+---
```
This file is a rework of this existing one -> https://posthog.com/docs/getting-started/identify-users

I've stripped out a lot, in the hopes of making things clearer and easier to follow. Let me know if I've missed anything important

Lior539 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,382 @@
+
+</MultiLanguage>
+
+<!-- ### Merging dangerously (TBD)
+
+1. `merge`, with no safeguards. We don't recommend using this in your code, but rather as a one-off manually for recovering from implementation problems. 
+   
+> **Warning:** Merging users with `merge` is irreversible! Be careful not to merge users who should not be merged together.
+
+For example using the posthog python library:
+
+```py
+posthog.alias('user-id', 'non-identified-id')
+
+posthog.capture('user-id', '$merge_dangerously', {'alias': 'second-user-id'})
+
+``` -->
````
I dont fully understand merge yet - would love to discuss on a huddle, and if it's even worth including. 

Also, I didn't find a `merge` function in the Node SDK. Is it only possible to call with `capture`?

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,382 @@
+
+---
+
+Identify enables you to to track your users across different platforms, and connect events from before and after they log in for the first time.
```
JavaScript (client) `identify` works a lot differently than server side `identify`. This definition works with the client side, not really the server side. [See this](https://posthog.com/tutorials/identifying-users-guide).

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+To set properties for a user, you can pass in a `properties` object in each `identify` call. Properties can be passed in either the `set` or `set_once` parameters.
+
+<MultiLanguage selector="tabs">
+
```
I need to update this code still

tiina303 replied:
jup sure - but more like if you already send an identify event you can set user props there, not that's the way to set user props

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+## Merging users
+
+It may happen that, due to implementation issues, the same user in your product has multiple users in PostHog associated with them. This occurs when, for the same user, you call `identify` with different distinct IDs. For example, if you used "email" as a distinct ID but then user changed their email. In these cases, you can use `merge` to merge multiple PostHog users into a single user.
+   
+> **Warning:** Merging users with `merge` is irreversible and has no safeguards! Be careful not to merge users who should not be merged together.
+> 
+> Due to the dangers, we don't recommend you merge users frequently, but rather as a one-off for recovering from implementation problems. We also encourage to consider whether merging users is strictly necessary for your data integrity, or a 'nice to have'.
```
```suggestion
It may happen that, due to implementation issues, the same user in your product has multiple users in PostHog. After the implementation has been fixed to no longer create duplicates the special `$merge_dangerously` event can be used to merge any users in PostHog.
   
> **Warning:** `$merge_dangerously` has no safeguards! Misusage could result in all users being merged into one rendering any analytics useless. Use it only as one-off for recovering from implementation problems. 
```

we don't need an example here - what they will see is duplicate persons in PostHog. 
Move this below considerations - could be part of the fixing issues you saw from ingestion warnings 🤷‍♀️

Lior539 replied:
good idea

tiina303 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,561 @@
+
+```ruby
+posthog.capture({
+    distinct_id: 'distinct id',
````
```suggestion
    distinct_id: 'distinct_id_of_user_to_merge_into',
```

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+</MultiLanguage>
+
+## Merging users
```
```suggestion
## Duplicate users in PostHog
```

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+  - `none`
+  - `null`
+  - `0`
+
```
- We do not allow merging from a "logged-in user"

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+- We process the event normally (it will be ingested and show up in the UI)
+- We refuse to merge users and an ingestion warning will be logged (see [ingestion warnings](/manual/data-management#ingestion-warnings) for more details).
+- The event will be only be tied to the user in the first passed `distinct_id`
```
```suggestion
- The event will be only be tied to the user associated with `distinct_id`
```

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+</MultiLanguage>
+
+### How are properties managed when merging users?
```
I dont think so, since in these docs we mention that alias is only for assigning a new id, and not for "identifying"

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+## How Identify is implemented in PostHog
+
+All events within PostHog are associated with a specific person, either an **anonymous user** (i.e., if `identify` has not been called) or an **identified user**.
```
Hmm Im not sure it's important to match the definitions 100%. Since we're recommending people use `identify` only (and not `alias`) for identifying users, it makes sense in this context.

tiina303 replied:
it's more about us reading the docs or anyone reading the code - it's confusing if we use the same name, but it means different things

Lior539 replied:
Fair point. I'll have a think on this

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+By using `identify`, you are associating the anonymous ID with the distinct ID that you have provided. This enables us to associate any past or future events with your user.
+
+This is helpful in scenarios such as analyzing onboarding flows, as you're able to see which actions and events a specific user triggered before they signed up.
```
Thanks. I dont think this line is needed actually, as we've already mentioned in the intro why identify is useful.

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+> Whenever possible, we suggest passing in all user properties you have available each time you call identify, as this will ensure that their user profile on PostHog is up to date.
+
+It is also possible to set user properties without using `identify` and by using `posthog.people` instead. This is helpful when you'd like to set properties for anonymous users (users who have not yet called `identify`).
```
good to know, thanks! I dont think this is needed then (since you can do the same using `capture`)

Lior539 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,561 @@
+posthog.people.set_once({ first_visited_URL: '/blog' })
+```
+
+### What is the difference between `set` and `set_once`?
````
Good to know. It's worth pointing out. How often does "ingestion time" occur?

> Maybe this page could be only about identify / merge (remove everything about setting user props from this one?).

Yeah, might be a good idea. The concepts are closely related though, and usually when you identify a user, you want to set their properties too. I'll have a think on this

tiina303 replied:
you can cross link the pages.
ingestion time vs event timestamp is relevant for historical imports or any time you're sending events out of order

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,561 @@
+
+</MultiLanguage>
+
+There are two requirements when assigning an `alias_id`:
+
+1. It cannot be associated with more than one `distinct_id`.
+2. The `alias_id` **must not** be equal to any other existing `distinct_id` i.e., it must be unique across all distinct IDs. For example:
```
Hmm I think the way I have written it is clearer, as having fewer terms is easier to understand ("logged in" user, or whatever we decide to call it)

tiina303 replied:
yes I propose to use "logged-in" user everywhere & not say "identified user" and this is where it would be helpful.

Lior539 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,561 @@
+
+<MultiLanguage selector="tabs">
+
+```node
+// For User A
+client.identify({
+    distinctId: 'distinct_id_one',
+    properties: {},
+})
+// For User B
+client.identify({
+    distinctId: 'distinct_id_two',
+    properties: {},
+})
+
+// ❌ Not possible. 
+// You cannot use distinct_id_two as an alias for distinct_id_one
+client.alias({
+    distinctId: 'distinct_id_one',
+    alias: 'distinct_id_two',
````
Yeah, thats much better!

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+---
+
+Identify enables you to to track your users across different devices, platforms (frontend and backend), and connect events from before and after they log in for the first time.
```
Identify doesn't really enable you to track users, `capture` tracks users. Identify is about adding or editing information to the users behind events.

You can track users without ever using `identify`

Lior539 replied:
I think the key phrase that's missing here is `the same user`

ivanagas commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,518 @@
+})
+```
+
+```javascript
````
It is important to call out that the client side works differently (connects with anonymous ID).

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+</MultiLanguage>
+
+> We don't recommend using emails as distinct IDs, as people changing their email will lead to extra complexity to maintaining data integrity. It is better to use database IDs or account IDs.
```
We use a lot of emails as distinct IDs, it works alongside database or account IDs.

Lior539 replied:
@Twixes any thoughts on this? IIRC, you recommended to a user on Slack to not use emails

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+> We don't recommend using emails as distinct IDs, as people changing their email will lead to extra complexity to maintaining data integrity. It is better to use database IDs or account IDs.
+
+You should call identify as soon as you're able to. Typically, this is every time your app loads for the first time, or directly after your users log in. This ensures that events sent during your users' sessions are correctly associated with them. 
```
Code snippet example?

Lior539 replied:
the snippet is above

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+You should call identify as soon as you're able to. Typically, this is every time your app loads for the first time, or directly after your users log in. This ensures that events sent during your users' sessions are correctly associated with them. 
+
+You can call identify multiple times for a user throughout the same session, **as long as you continue to pass the same distinct_id**.
```
I don't understand why you would want to do this?

Lior539 replied:
It's not that you'd want to do this, but rather providing assurance that its fine to do this and wont cause problems. But perhaps we don't need this line and it can be removed.

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+All events within PostHog are associated with a specific person, either an **anonymous user** (i.e., if `identify` has not been called) or an **identified user**.
+
+When a user starts browsing your website or app, they'll be automatically assigned an **anonymous ID**, which is then stored locally and allows us to track anonymous users – even across different sessions. This anonymous ID is created using the user's device ID, and typically looks like this: `17b845b08de74-033c497ed2753c-35667c03-1fa400-17b845b08dfd55`.
```
A lot of this statement depends. If you have the JS SDK or snippet, then they are automatically assigned an anon ID, but not with backend SDKs. 

Also, there are a bunch of configuration options for initializations that means it might not always be stored locally (for example, persistence: false), and can't track across sessions.

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+When a user starts browsing your website or app, they'll be automatically assigned an **anonymous ID**, which is then stored locally and allows us to track anonymous users – even across different sessions. This anonymous ID is created using the user's device ID, and typically looks like this: `17b845b08de74-033c497ed2753c-35667c03-1fa400-17b845b08dfd55`.
+
+By using `identify`, you are associating the anonymous ID with the distinct ID that you have provided. This enables us to associate any past or future events with your user.
```
On the frontend yes, on the backend no, unless something has changed?

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+## Considerations
+
+Identifying users is a powerful feature, but it also has the potential to create problems if misused. 
```
```suggestion
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+Identifying users is a powerful feature, but it also has the potential to create problems if misused. 
+
+An important mistake to avoid is using non-unique distinct IDs to identify users. Two common ways in which this can happen are:
```
```suggestion
Avoid using non-unique distinct IDs to identify users. Two common ways in which this can happen are:
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+- Your logic for generating IDs does not generate sufficiently strong IDs and you can end up with a clash where 2 users have the same ID.
+
+- There's a bug, typo, or mistake in your code leading to most or all users being identified with generic IDs like `null`, `true`, or `distinctId`.
+
```
Not handling a change in users properly, like logging out.

ivanagas reviewed (commented):
More notes

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,518 @@
+
+</MultiLanguage>
+
+## Considerations
```
Good call

Lior539 commented on contents/docs/getting-started/user-properties.mdx:
```diff
@@ -53,25 +53,24 @@ posthog.capture(
 
 User property values can be strings, booleans, numbers, objects, or arrays. Note that for objects and arrays, you'll need to use the [Property Flattener app](apps/property-flattener) in order to flatten nested properties and make them accessible through filters in the PostHog app.
```
I removed the user properties section from the identify docs. However, since I made some small improvements to the text, I'm including these updates in this PR

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,466 @@
+
+In your frontend, all events within PostHog are associated with a specific person:
+1. Either an **anonymous user** – i.e., if `identify` has not been called.
+2. Or a **recognized user** – i.e., `identify` has been called..
```
I feel slightly uncomfortable about introducing the term "recognized user", but maybe its okay since it's only purpose is to contrast it with an anonymous user

Lior539 commented:
Thank you both @tiina303 + @ivanagas for the constructive comments. I think it's in a much better place, and requesting what I hope is the final review 🤞🏼

Note that there's still a bunch more work to do with Identify+alis in the docs, like unifying+deduplicating the docs across the SDK docs and the Getting started section. I'll do this in a separate PR, as this PR is already quite large

tiina303 reviewed (changes requested):
sorry ... for the many change requests, but on a positive note I think I just came up with an idea how to make this doc clear for users.

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,466 @@
+
+## Alias
+
+Sometimes, you may want to assign multiple distinct IDs to a single user. This is helpful in scenarios where your primary distinct ID may be inaccessible. For example, if a distinct ID which is typically used on the frontend is not available in certain parts of your backend code. In this case, you can use `alias` to assign another distinct ID to the same user:
+
```
While sub-system is technically more correct, since this is just an example with the purpose of illustrating a scenario, I think it's simpler and easier to understand if we use backend or frontend

Lior539 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,466 @@
+```
+
+```python
+posthog.alias('alias_id', 'distinct_id');
````
good catch!

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,466 @@
+
+</MultiLanguage>
+
+There are two requirements when assigning an `alias_id`:
+
+1. It cannot be associated with more than one `distinct_id`.
+2. The `alias_id` **must not** be the `distinct_id` of any other user. For example:
```
What's missing?

tiina303 replied:
if you have called `$identify(distinct_id, ...)` anywhere even in the backend and without another id, then you can't use that distinct_id as alias_id anymore. The order in which you provide id's is important. I think the previous example was better we had, it's more clear how someone would actually run into it
```
identify('distinct_id')
alias(distinct_id = 'new_id', alias_id = 'distinct_id')  // doesn't work
alias(distinct_id = 'distinct_id;, alias_id = 'new_id') // does work assuming 'new_id' isn't used anywhere.
```

Lior539 replied:
Hmm interesting. I agree the previous example was clearer, but I removed it since we are encouraging users not to call identify in the backend (since it doesnt really make sense to use it in the backend), so this is why I removed `identify` in these examples.

I dont feel too strongly about this though. Happy to change it back if you prefer. What do you think?

tiina303 replied:
you could just say imagine in the past we've called this on a separate session in front of the examples.

tiina303 replied:
The actual restriction (that I'm about to change the code to have) is if you've used this distinct_id as distinct_id for ('$identify', '$create_alias' or '$merge_dangerously') then you can'd use it as anon nor alias ID anymore

Lior539 replied:
Hmm I updated the example to be similar to what it was previously. What do you think? 
I think we should keep it in this section to help paint a picture of how to use alias

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,466 @@
+
+Identify enables you to to track the same user across different devices, platforms (frontend and backend), and connect events from before and after they log in for the first time.
+
+Identifying users is done using the `identify` method, together with a unique ID. It is available in both our frontend and backend SDKs, as well as our [POST API](/docs/api/post-only-endpoints).
```
I dont follow 100%. I think I have a knowledge gap in how identify works on the backend. Can we have a huddle to discuss?

Lior539 commented:
I've done a complete rework based on the feedback. There is now a much stronger emphasis on anonymous users and using identify on the frontend (and not recommending to use it on the backend).

I think its in a much better state. Hopefully we are near the end :)

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,408 @@
+
+If you've accidentally linked an incorrect anonymous ID with a specific distinct ID, or you've made a mistake when merging users, it's possible to split their combined user back into separate users. You can do this in the PostHog app by navigating to the user you'd like split, and then clicking "Split IDs" in the top right corner. 
+
+> **Warning:** This may change the numbers in your charts, even historically. We also cannot guarantee that historical data will be assigned to the correct user after the split. 
```
@tiina303 anything worth adding/changing about this warning

tiina303 replied:
I'd just say it this way - note that it's not something that's used frequently - this way we are not limited in any way in the future.
```suggestion
> **Warning:** Going forward the distinctIDs are treated as separate users, but what happens to historical data is undefined (the distinctIDs could be split up to separate users or kept together or partially split up).
```

Lior539 replied:
```suggestion
> **Warning:** This will treat the distinct IDs as separate users for future events. However, there is no guarantee as to how past events will be treated – they may be be considered separate users, or be considered a single user for some or all events.
```

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,408 @@
+
+</MultiLanguage>
+
+If we encounter an `identify` or `alias` event with one of the above problems, the following will happen:
+
+1. We process the event normally (it will be ingested and show up in the UI)
+2. Merging users will be refused and an ingestion warning will be logged (see [ingestion warnings](/manual/data-management#ingestion-warnings) for more details).
+3. The event will be only be tied to the user associated with `distinct_id`
```
this is true for when we encounter null etc invalid distinctIDs - the considerations and what will happen should be together and have this in the end. Move both to troubleshooting/FAQ section. into the I have dups section

tiina303 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,408 @@
+When an anonymous user is identified as `User A`, all the properties of the anonymous user are added to `User A`. If there is a conflict, the properties of `User A` are prioritized over the anonymous user. For example:
+
+```js
+client.identify({
+    distinctId: 'user_A',
+    properties: {
+        name: 'User A',
+        location: 'London',
+        timezone: 'GMT',
+    },
+})
+
+client.reset()
````
I'd use for simplicity, so they don't need to go back understanding identify and reset
```suggestion
/* User A have properties:
{ 
        name: 'User A',
        location: 'London',
        timezone: 'GMT',
}
*/
```

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,408 @@
+   
+> **Warning:** Merging users with `merge_dangerously` is irreversible and has no safeguards! Be careful not to merge users who should not be merged together.
+> 
+> Due to the dangers, we don't recommend you merge users frequently, but rather as a one-off for recovering from implementation problems. We also encourage you to consider whether merging users is strictly necessary for your data integrity, or a 'nice to have'.
```
why would we say that?
```suggestion
> Due to the dangers, we don't recommend you merge users frequently, but rather as a one-off for recovering from implementation problems.
```

tiina303 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,408 @@
+
+</MultiLanguage>
+
+#### How are properties managed when merging users?
+
+
+When a `User B` is merged into another `User A`, all the properties of the `User B` are added to `User A`. If there is a conflict, the properties of `User A` are prioritized over `User B`. For example:
+```js
+client.identify({
+    distinctId: 'user_A',
+    properties: {
+        location: 'London',
+        name: 'User A',
+    },
+})
+
+client.identify({
+    distinctId: 'user_B',
+    properties: {
+        name: 'User B',
+        location: 'Rome',
+        phone: '0800-POSTHOG',
+    },
+})
+
+posthog.alias(
+  'user_A', // distinct_id
+  'user_B'  // alias_id
+)
+
+/* User B has merged into User A. User A will now have properties:
+{ 
+  name: 'User A',
+  location: 'London', 
+  phone: '0800-POSTHOG', 
+}
+*/
+```
````
This doesn't work for several reasons:
1. alias_id you use is already identified user (whenever you use a distinct_id as distinct_id in identify call it is now `is_identified` even if an actual merge didn't happen)
2. calling identify in a row the second call would merge the users already (even if you said imagine different browsers it's still simpler to just call out the properties state).
3. This shows a bad use case of merging 2 users based on properties looking like they belong to different users

I suggest we call out the likely reasonable use case (rather than merging users, also move it to be after identify in the FAQ so it's more logical ordering. 
Furthermore let's not use js here as the alias call language for clarity.

```suggestion
#### How are properties managed when assigning multiple distinct IDs to the same user?

When you use alias or `$merge_dangerously` to merge 2 distinctIDs that represent the same user, then sometimes both of those PostHog users have properties. When a `distinct_id_B` is merged into another `distinct_id_A`, all the properties of the `distinct_id_B` are added to `distinct_id_A`. If there is a conflict, the properties of `distinct_id_A` are prioritized over `distinct_id_B`. For example:
```node
/* distinct_id_A user in PostHog has these properties
{
        location: 'London',
        name: 'Max the Hedgehog',
        favorite_food: 'pineapple'
},
distinct_id_B user in PostHog has these properties
{
        phone: '0800-POSTHOG',
        favorite_food: 'pizza'
}
*/
client.alias({
    distinctId: 'distinct_id_A',
    alias: 'distinct_id_B',  // if both users have been identified or aliased in the past use `$merge_dangerously`
})

/* The resulting user will now have properties:
{ 
  name: 'Max the Hedgehog',
  location: 'London', 
  phone: '0800-POSTHOG', 
  favorite_food: 'pineapple'
}
*/
```
```

Lior539 replied:
Hmm you're right, this is not a good example.  We should use `merge_dangerously` code instead of `alias` for this example.

> also move it to be after identify in the FAQ so it's more logical ordering.

Im not sure what you mean by this.

> #### How are properties managed when assigning multiple distinct IDs to the same user?

I think its simpler to separate the `alias` merge from the `$merge_dangerously` merge, since the `alias` merge applies to anonymous users only right?

tiina303 replied:
for properties merging it doesn't matter if it was anonymous user or not

tiina303 replied:
> also move it to be after identify in the FAQ so it's more logical ordering.

FAQ:
## how anonymous properties are merged
## how added distinct ID properties are merged (these are the same actually)
## other sections

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,408 @@
+
+### How to split a merged user back into separate users
+
+If you've accidentally linked an incorrect anonymous ID with a specific distinct ID, or you've made a mistake when merging users, it's possible to split their combined user back into separate users. You can do this in the PostHog app by navigating to the user you'd like split, and then clicking "Split IDs" in the top right corner. 
```
Not only anonymous ID problem
```suggestion
If you've accidentally linked distinct IDs together that represent different users, or you've made a mistake when merging users, it's possible to split their combined user back into separate users. You can do this in the PostHog app by navigating to the user you'd like split, and then clicking "Split IDs" in the top right corner. 
```

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,408 @@
+
+</MultiLanguage>
+
+If you _also_ want to reset the `device_id` so that the device will be considered a new device in future events, you can pass `true` as an argument:
```
Yeah. I'm not sure what a use case of this is tbh. I suspect people will know if they want to do this or not, so I suppose we dont need an example

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,408 @@
+client.reset()
+
+// as an anonymous user
+posthog.people.set({ 
+    name: 'Anonymous User',
+    phone: '0800-POSTHOG' 
+})
```
I should update this to capture

Lior539 commented:
Changes:
* Added callout when merging users with Alias
* Moved "What happens if you call `identify` or `alias` with invalid inputs?" to FAQ
* updated merge dangerously example

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,462 @@
+
+</MultiLanguage>
+
+> Note that when calling `alias` in the frontend SDKs, if you have set any properties onto the anonymous user, they will be merged into the user with `distinct_id`. For more details, see the FAQ on [how properties are managed when identifying anonymous users](/docs/data/identify#how-are-properties-managed-when-identifying-anonymous-users).
```
this is new

tiina303 replied:
it's exactly the same as #### How are properties managed when merging users? why not link there?

tiina303 replied:
it's exactly the same as #### How are properties managed when merging users? why not link there?

Lior539 replied:
I think it's good to keep it as its with the rest of the context with alias.

Re having two section "#### How are properties managed when merging users?' and "####how-are-properties-managed-when-identifying-anonymous-users": While they are technically the same, it's easier to explain them as two different sections, since they apply to separate scenarios (one is frontend+anonymous users, while the other scenario is around merging in the backend)

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,462 @@
+
+1. We process the event normally (it will be ingested and show up in the UI).
+2. Merging users will be refused and an ingestion warning will be logged (see [ingestion warnings](/manual/data-management#ingestion-warnings) for more details).
+3. In the case of `alias`, the event will be only be tied to the user associated with `distinct_id`.
```
This was a different section before, and has now been moved to the FAQ

Lior539 commented on contents/docs/data/identify.mdx:
````diff
@@ -0,0 +1,462 @@
+
+When a `User B` is merged into another `User A`, all the properties of the `User B` are added to `User A`. If there is a conflict, the properties of `User A` are prioritized over `User B`. For example:
+
+```node
+/* Assume User A has the properties:
+{ 
+        name: 'User A',
+        location: 'London',
+}
+*/
+
+/* Assume User B has the properties:
+{ 
+        name: 'User B',
+        location: 'Rome',
+        phone: '0800-POSTHOG',
+}
+*/
+
+client.capture({
+    distinctId: 'distinct_id_of_user_A',
+    event: '$merge_dangerously',
+    properties: {
+        alias: 'distinct_id_of_user_B',
+    },
+})
+
+/* User B has merged into User A. The resulting user will now have properties:
+{ 
+  name: 'User A',
+  location: 'London', 
+  phone: '0800-POSTHOG', 
+}
+*/
+```
````
This example has been updated

tiina303 replied:
that's fine though I'm not sure it provides good incentives to merge users with different names

tiina303 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,462 @@
+If two users have the same distinct ID, their data will be merged and they will be considered as one user in PostHog. Two common ways in which this can happen are:
+
+- Your logic for generating IDs does not generate sufficiently strong IDs and you can end up with a clash where 2 users have the same ID.
+- There's a bug, typo, or mistake in your code leading to most or all users being identified with generic IDs like `null`, `true`, or `distinctId`.
+
+PostHog also has a few built-in protections to stop the most common distinct ID mistakes:
+
+- We do not allow identifying users with empty space strings of any length – e.g.,`' '`, `'       '`, etc.
+- We do not allow identifying users with the following strings (case insensitive):
+    - `anonymous`
+    - `guest`
+    - `distinctid`
+    - `distinct_id`
+    - `id`
+    - `not_authenticated`
+    - `email`
+    - `undefined`
+    - `true`
+    - `false`
+
+- We do not allow identifying users with the following strings (case sensitive):
+  - `[object Object]` 
+  - `NaN`
+  - `None`
+  - `none`
+  - `null`
+  - `0`
+
+- We do not allow identifying a user that has already been identified with a different distinct ID. For example:
+
+<MultiLanguage selector="tabs">
```
This is too much detail for the main section - move it to troubleshooting / FAQ

Lior539 replied:
hmm I see your point, but this feels like more of a nit. It fits in nicely to this section, I would prefer to leave here

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+---
+Linking events to specific users enables you to gain full insights as to how they're using your product across different sessions, devices, and platforms. 
+
+This is straight forward to do when [capturing backend events](/docs/getting-started/send-events#3-capture-backend-events), as you associate events to a specific user using `distinct_id`, which is a required argument. However, on the [frontend]((/docs/getting-started/send-events#sending-custom-properties-on-an-event)), this is not a required argument and thus events can be submitted anonymously.  ...
```
```suggestion
This is straight forward to do when [capturing backend events](/docs/getting-started/send-events#3-capture-backend-events), as you associate events to a specific user using `distinct_id`, which is a required argument. However, on the [frontend]((/docs/getting-started/send-events#sending-custom-properties-on-an-event)), a `distinct_id` is not a required argument and thus events can be submitted anonymously. To link events from anonymous to specific users, you need to use `identify`:
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+## How `identify` works
+
+When a user starts browsing your website or app, they'll be automatically assigned an **anonymous ID**, which is then stored locally. This allows us to track anonymous users – even across different sessions.
```
```suggestion
When a user starts browsing your website or app, PostHog automatically assigns them an **anonymous ID**, which is stored locally. This enables us to track anonymous users – even across different sessions.
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+When a user starts browsing your website or app, they'll be automatically assigned an **anonymous ID**, which is then stored locally. This allows us to track anonymous users – even across different sessions.
+
+> Note that depending on your [persistence configuration](/docs/libraries/js#persistence), the anonymous ID may not be stored, and thus regenerated across sessions.
```
```suggestion
> **Note:** depending on your [persistence configuration](/docs/libraries/js#persistence), the anonymous ID may not be stored, and thus regenerated across sessions.
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+> Note that depending on your [persistence configuration](/docs/libraries/js#persistence), the anonymous ID may not be stored, and thus regenerated across sessions.
+
+By calling `identify` with a `distinct_id` of your choice (usually the user's ID in your database, or their email), you are linking the anonymous ID and distinct ID together.
```
```suggestion
By calling `identify` with a `distinct_id` of your choice (usually the user's ID in your database, or their email), you link the anonymous ID and distinct ID together.
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+> **A note on using `identify` in the backend:**
+>
+> Although it is technically possible to call `identify` using our backend SDKs, it is typically most useful in frontends. This is because there is no concept of anonymous sessions in the backend SDKs.
```
```suggestion
> Although it is technically possible to call `identify` using our backend SDKs, it is typically most useful in frontends. This is because there is no concept of anonymous sessions in the backend SDKs. `identify` on the backend is used to update user information.
```

Lior539 replied:
I would prefer to avoid saying this, as `capture` is the preferred method to do this.

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+### 1. Call `identify` as soon as you're able to
+
+In your frontend, you should call identify as soon as you're able to. Typically, this is every time your app loads for the first time, or directly after your users log in. This ensures that events sent during your users' sessions are correctly associated with them. 
```
```suggestion
In your frontend, you should call `identify` as soon as you're able to. Typically, this is every time your app loads for the first time, or directly after your users log in. This ensures that events sent during your users' sessions are correctly associated with them. 
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+You only need to call `identify` once per session.
+
+### 2. Use unique strings for Distinct IDs
```
```suggestion
### 2. Use unique strings for distinct IDs
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+### 2. Use unique strings for Distinct IDs
+
+If two users have the same distinct ID, their data will be merged and they will be considered as one user in PostHog. Two common ways in which this can happen are:
```
```suggestion
If two users have the same distinct ID, their data will be merged and they will be considered as one user in PostHog. Two common ways this can happen are:
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+- Your logic for generating IDs does not generate sufficiently strong IDs and you can end up with a clash where 2 users have the same ID.
+- There's a bug, typo, or mistake in your code leading to most or all users being identified with generic IDs like `null`, `true`, or `distinctId`.
+
+PostHog also has a few built-in protections to stop the most common distinct ID mistakes:
```
```suggestion
PostHog also has built-in protections to stop the most common distinct ID mistakes:
```

ivanagas commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,505 @@
+
+If a user logs out on your frontend, you should call `reset` to unlink any future events made on that device with that user. 
+
+This is important if your users are sharing a computer, as otherwise all of those users will be grouped together into a single user due to shared cookies between sessions. **We strongly recommend you do this on logout even if you don't expect users to share a computer.**
```
```suggestion
This is important if your users are sharing a computer, as otherwise all of those users are grouped together into a single user due to shared cookies between sessions. **We strongly recommend you call `reset` on logout even if you don't expect users to share a computer.**
```

ivanagas reviewed (approved):
Looks good once you sort the alias stuff out 😄

tiina303 reviewed (approved):
Re-iterating some comments of mine:
1. PostHog also has built-in protections to stop merging users <- these restrictions are the same for alias and identify (the restriction around illegal IDs and alias_id/anon_id need to not have been used as distinct_id in any of the functions in the past). I'd move this into FAQ "What happens if you call `identify`, `alias`, or `merge_dangerously` with invalid inputs?" & link to it from above.
2. Reorder Troubleshooting (1) properties merged with identify (2) properties merged with alias <- because those two are the same and very related. (3) What happens if you call `identify`, `alias`, or `merge_dangerously` with invalid inputs? (4) dup users (5) split users <- these later too are related, make sense to have them near each other too.

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,477 @@
+    {} 
+    {}, 
+);
+posthog.identify(
+    'distinct_id_two',
+    {} 
+    {}, 
+);
+// ❌ Not possible, since we already identified this user with "distinct_id_one"
+// so we cannot identify them again with a different distinct ID "distinct_id_two"
```
Thanks! Re" all future events will be tied to distinct_id_two, I dont think we need to add this detail to this docs

tiina303 replied:
I wouldn't, see the general comment too

Lior539 commented on contents/docs/data/identify.mdx:
```diff
@@ -0,0 +1,477 @@
+- Your logic for generating IDs does not generate sufficiently strong IDs and you can end up with a clash where 2 users have the same ID.
+- There's a bug, typo, or mistake in your code leading to most or all users being identified with generic IDs like `null`, `true`, or `distinctId`.
+
+PostHog also has built-in protections to stop the most common distinct ID mistakes:
```
Thanks! I thought about a way to include alias and merge dangerously here as well, but left it out as a way to avoid talking about them before we've described what they are.

I did also think about moving this to end of the doc to the FAQ, but I think it fits in better with this context

Lior539 commented:
> Reorder Troubleshooting (1) properties merged with identify (2) properties merged with alias <- because those two are the same and very related. (3) What happens if you call identify, alias, or merge_dangerously with invalid inputs? (4) dup users (5) split users <- these later too are related, make sense to have them near each other too.

Thanks. I will reorder similar to what you have suggested, but `(4)  dup users` needs to be before `2)properties merged with alias`, since `2) mentions` merge_dangerously` but `(4)  dup users`  is the question that introduces it. So order will be:

(4) dup users 
(5) split users 
 (1) properties merged with identify 
(2) properties merged with alias <- because those two are the same and very related. 
(3) What happens if you call identify, alias, or merge_dangerously with invalid inputs?
